### PR TITLE
Rework CI

### PR DESCRIPTION
**Pull request by cwfitzgerald**
**Checklist**

- [x] Run `cargo clippy`.
- [x] Run `RUSTFLAGS=--cfg=web_sys_unstable_apis cargo clippy --target wasm32-unknown-unknown` if applicable.
- [x] Add change to CHANGELOG.md. See simple instructions inside file.

First pass iterative improvement on CI - this cleans up the workflow a lot, switches to swiftshader, and updates some uncaught broken tests.

### CODE REVIEW

**Comment by cwfitzgerald** [comment]
@jimblandy I either need the permissions, or I need you to change the required successes to the new CI names

File: wgpu-hal/src/dx12/device.rs [hunk 1]
```diff
@@ -1260,10 +1260,10 @@ impl crate::Device<super::Api> for super::Device {
     }
     unsafe fn destroy_bind_group(&self, group: super::BindGroup) {
         if let Some(dual) = group.handle_views {
-            let _ = self.shared.heap_views.free_slice(dual);
+            self.shared.heap_views.free_slice(dual);
         }
         if let Some(dual) = group.handle_samplers {
-            let _ = self.shared.heap_samplers.free_slice(dual);
+            self.shared.heap_samplers.free_slice(dual);
```
Comment (jimblandy) [inline wgpu-hal/src/dx12/device.rs:1266]: This isn't really related to this PR, is it?

File: wgpu/examples/water/main.rs [hunk 2]
```diff
@@ -826,9 +826,8 @@ fn water() {
         height: 768,
         optional_features: wgpu::Features::default(),
         base_test_parameters: framework::test_common::TestParameters::default()
-            .downlevel_flags(wgpu::DownlevelFlags::READ_ONLY_DEPTH_STENCIL)
-            .specific_failure(Some(wgpu::Backends::DX12), None, Some("Basic"), false), // WARP has a bug https://github.com/gfx-rs/wgpu/issues/1730
+            .downlevel_flags(wgpu::DownlevelFlags::READ_ONLY_DEPTH_STENCIL),
         tolerance: 5,
-        max_outliers: 470, // bounded by DX12, then AMD Radeon Polaris12 on vk linux
+        max_outliers: 1693, // bounded by swiftshader
```
Comment (jimblandy) [inline wgpu/examples/water/main.rs:831]: I see a lot of outlier counts being increased. Are there others that could be decreased? The point at which we switch back ends would be a natural time to reevaluate those.

File: wgpu/tests/common/mod.rs [hunk 3]
```diff
@@ -278,7 +278,10 @@ pub fn initialize_test(parameters: TestParameters, test_function: impl FnOnce(Te
     }
 
     let panicked = catch_unwind(AssertUnwindSafe(|| test_function(context))).is_err();
+    #[cfg(any(not(target_arch = "wasm32"), target_os = "emscripten"))]
     let canary_set = hal::VALIDATION_CANARY.get_and_reset();
+    #[cfg(not(any(not(target_arch = "wasm32"), target_os = "emscripten")))]
```
Comment (jimblandy) [inline wgpu/tests/common/mod.rs:283]: my kingdom for an else

**Review by jimblandy — APPROVED** [review]
Looks good to me. Some minor questions.

File: wgpu-hal/src/dx12/device.rs (hunk 1 above)
Comment (cwfitzgerald) [reply wgpu-hal/src/dx12/device.rs:1266]: This is the outstanding lint that wasn't caught by the previous edition of the CI, so I had to fix it to get this CI to pass

File: wgpu/tests/common/mod.rs (hunk 3 above)
Comment (cwfitzgerald) [reply wgpu/tests/common/mod.rs:283]: In all seriousness I should check if cfg_if is in our deps.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -126,118 +93,221 @@ jobs:
...
-          # build for WebGL
-          cargo ${{matrix.tool}} --target ${{ matrix.target }} -p wgpu --features webgl
+          for backend in ${{ matrix.backends }}; do
+            echo "======= NATIVE TESTS $backend ======";
+            WGPU_BACKEND=$backend cargo llvm-cov nextest -p wgpu -p wgpu-types -p wgpu-hal -p wgpu-core -p player --no-fail-fast --no-report
+          done
 
-          # build docs
-          cargo doc --target ${{ matrix.target }} -p wgpu --no-deps
+          cargo llvm-cov report --lcov --output-path lcov.info
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply .github/workflows/ci.yml:253]: Refactor residue :)

File: wgpu/examples/water/main.rs (hunk 2 above)
Comment (cwfitzgerald) [reply wgpu/examples/water/main.rs:831]: I want to re-look at these in a separate pass - these already have limited usefulness.

File: wgpu/tests/common/mod.rs (hunk 3 above)
Comment (cwfitzgerald) [reply wgpu/tests/common/mod.rs]: We did have it in our deps already, so replaced this with cfg-if

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -126,118 +93,221 @@ jobs:
...
 
           # vulkan sdk
           wget -qO - https://packages.lunarg.com/lunarg-signing-key-pub.asc | sudo apt-key add -
           sudo wget -qO /etc/apt/sources.list.d/lunarg-vulkan-focal.list https://packages.lunarg.com/vulkan/lunarg-vulkan-focal.list
 
           sudo apt-get update
-          sudo apt install -y libegl1-mesa libgl1-mesa-dri libxcb-xfixes0-dev mesa-vulkan-drivers vulkan-sdk
+          sudo apt install -y libegl1-mesa libgl1-mesa-dri libxcb-xfixes0-dev vulkan-sdk
 
-      # We enable line numbers for panics, but that's it
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply .github/workflows/ci.yml:170]: Added comment explaining rationale.

**Pull request merged**